README.md
# toastr2

[![Build Status](https://travis-ci.com/JPeer264/toastr2.svg?branch=master)](https://travis-ci.com/JPeer264/toastr2)
[![Coverage Status](https://coveralls.io/repos/github/JPeer264/toastr2/badge.svg?branch=master)](https://coveralls.io/github/JPeer264/toastr2?branch=master)

> Note: this is a fork of [toastr](https://github.com/CodeSeven/toastr). This library is made to use it without jQuery and with Typescript

**toastr** is a Javascript library for non-blocking notifications The goal is to create a simple core library that can be customized and extended.

## Demo
- Demo can be found at https://jpeer264.github.io/toastr2/demo.html

## Install

#### [npm](https://www.npmjs.com/package/toastr2)
```
npm install toastr2
```

#### [yarn](https://yarnpkg.com/en/package/toastr2)
```
yarn add toastr2
```

## Breaking Changes

#### Animation Changes
The following animations options have been deprecated and should be replaced:

 - Replace `options.fadeIn` with `options.showDuration`
 - Replace `options.onFadeIn` with `options.onShown`
 - Replace `options.fadeOut` with `options.hideDuration`
 - Replace `options.onFadeOut` with `options.onHidden`

## Quick Start

### 2 Easy Steps
For other API calls, see the [demo](https://jpeer264.github.io/toastr2/demo.html).

1. Import toastr2 (js and css)
	```js
	import Toastr from 'toastr2';

	import 'toastr2/dist/toastr.min.css';

	const toastr = new Toastr();
	```

2. use toastr to display a toast for info, success, warning or error
	```js
	// Display an info toast with no title
	toastr.info('Are you the 6 fingered man?')
	```

### Other Options
```js
// Display a warning toast, with no title
toastr.warning('My name is Inigo Montoya. You killed my father, prepare to die!')

// Display a success toast, with a title
toastr.success('Have fun storming the castle!', 'Miracle Max Says')

// Display an error toast, with a title
toastr.error('I do not think that word means what you think it means.', 'Inconceivable!')

// Immediately remove current toasts without using animation
toastr.remove()

// Remove current toasts using animation
toastr.clear()

// Override global options
toastr.success('We do have the Kapua suite available.', 'Turtle Bay Resort', {timeOut: 5000})
```

### Escape HTML characters
In case you want to escape HTML characters in title and message

	toastr.options.escapeHtml = true;

### Close Button
Optionally enable a close button
```js
toastr.options.closeButton = true;
```

Optionally override the close button's HTML.

```js
toastr.options.closeHtml = '<button><i class="icon-off"></i></button>';
```

You can also override the CSS/LESS for `#toast-container .toast-close-button`

Optionally override the hide animation when the close button is clicked (falls back to hide configuration).
```js
toastr.options.closeMethod = 'fadeOut';
toastr.options.closeDuration = 300;
toastr.options.closeEasing = 'swing';
```

### Display Sequence
Show newest toast at bottom (top is default)
```js
toastr.options.newestOnTop = false;
```

### Callbacks
```js
// Define a callback for when the toast is shown/hidden/clicked
toastr.options.onShown = () => console.log('hello');
toastr.options.onHidden = () => console.log('goodbye');
toastr.options.onclick = () => console.log('clicked');
toastr.options.onCloseClick = () => console.log('close button clicked');
```

### Animation Options
Toastr will supply default animations, so you do not have to provide any of these settings. However you have the option to override the animations if you like.

#### Easings
Optionally override the animation easing to show or hide the toasts. Default is swing. swing and linear are built into jQuery.
```js
toastr.options.showEasing = 'swing';
toastr.options.hideEasing = 'linear';
toastr.options.closeEasing = 'linear';
```

Using the jQuery Easing plugin (http://www.gsgd.co.uk/sandbox/jquery/easing/)
```js
toastr.options.showEasing = 'easeOutBounce';
toastr.options.hideEasing = 'easeInBack';
toastr.options.closeEasing = 'easeInBack';
```

#### Animation Method
Use the jQuery show/hide method of your choice. These default to fadeIn/fadeOut. The methods fadeIn/fadeOut, slideDown/slideUp, and show/hide are built into jQuery.
```js
toastr.options.showMethod = 'slideDown';
toastr.options.hideMethod = 'slideUp';
toastr.options.closeMethod = 'slideUp';
```

### Prevent Duplicates
Rather than having identical toasts stack, set the preventDuplicates property to true. Duplicates are matched to the previous toast based on their message content.
```js
toastr.options.preventDuplicates = true;
```

### Timeouts
Control how toastr interacts with users by setting timeouts appropriately.
```js
toastr.options.timeOut = 30; // How long the toast will display without user interaction
toastr.options.extendedTimeOut = 60; // How long the toast will display after a user hovers over it
```

### Prevent from Auto Hiding
To prevent toastr from closing based on the timeouts, set the `timeOut` and `extendedTimeOut` options to `0`. The toastr will persist until selected.

```js
toastr.options.timeOut = 0;
toastr.options.extendedTimeOut = 0;
```

### Progress Bar
Visually indicate how long before a toast expires.
```js
toastr.options.progressBar = true;
```

### rtl
Flip the toastr to be displayed properly for right-to-left languages.
```js
toastr.options.rtl = true;
```

## Building Toastr

To build the minified and css versions of Toastr you will need [node](http://nodejs.org) installed. (Use Homebrew or Chocolatey.)

```
npm install
```

At this point the dependencies have been installed and you can build Toastr

- Run the analytics `npm run lint`
- Run the test `npm test`
- Run the build `npm run build`

## Contributing

For a pull request to be considered it must resolve a bug, or add a feature which is beneficial to a large audience.

Pull requests must pass existing unit tests, CI processes, and add additional tests to indicate successful operation of a new feature, or the resolution of an identified bug.

Requests must be made against the `develop` branch. Pull requests submitted against the `master` branch will not be considered.

All pull requests are subject to approval by the repository owners, who have sole discretion over acceptance or denial.

## Authors
**John Papa**

+ [http://twitter.com/John_Papa](http://twitter.com/John_Papa)

**Tim Ferrell**

+ [http://twitter.com/ferrell_tim](http://twitter.com/ferrell_tim)

**Hans Fjällemark**

+ [http://twitter.com/hfjallemark](http://twitter.com/hfjallemark)

## Credits
Inspired by https://github.com/Srirangan/notifer.js/.

## Copyright
Copyright © 2012-2015

## License
toastr is under MIT license - http://www.opensource.org/licenses/mit-license.php


CHANGELOG.md
3.0.0-alpha.18 - December, 10 2020

* 49d351f Chore: revert to lowercase .css file | copy scss to dist for sourcemaps (JPeer264)

3.0.0-alpha.17 - December, 01 2020

* e9b0b9e Chore: support case-sensitive platforms (JPeer264)

3.0.0-alpha.16 - December, 01 2020

* 72a257f Chore: include src for npmignore for types and sourcemaps (JPeer264)

3.0.0-alpha.15 - April, 13 2020

* 3915ff1 Docs: forgot ' in usage import (JPeer264)
* bda5794 Chore: ignore more folders for npm to reduce download size (JPeer264)

3.0.0-alpha.14 - March, 18 2020

* dc5d8f4 Test: add comment on skipped test (JPeer264)
* 16a79e2 Test: add tests for remove and clear (JPeer264)
* e4b3bba Test: add tests for debug (JPeer264)
* e1c10ae Chore: update typings (JPeer264)
* 34322dc Chore: update types for functions (JPeer264)
* 3a82472 Chore: fix coverage (JPeer264)
* 19956e0 Docs: remove cdns (JPeer264)
* 3802d60 Chore: add coveralls and badges (JPeer264)
* 177f1e2 Docs: update demo with npm script (JPeer264)

3.0.0-alpha.13 - March, 04 2020

* 2be561f Fix license in package.json (#4) (Matthias Schiebel)
* c37143d Chore: update node-sass for node v12 (JPeer264)
* 97b8b72 Chore: update readme with new docs (JPeer264)
* ae204c8 Chore: drop node v8 and use v10 and v12 (JPeer264)
* 99a83bf Chore: use .flat shim for node v10 (JPeer264)

3.0.0-alpha.12 - February, 23 2020

* 3e2bef0 Fix: add more cases with string[] to add classes (ref #3) (JPeer264)

3.0.0-alpha.11 - February, 22 2020

* 0d66761 Test: remove old test setup files (JPeer264)
* 34212fd Test: refactor positioning tests (JPeer264)
* 2c24295 Test: refactor subscription and appearance (JPeer264)
* 6c44622 Test: finish event refactoring (JPeer264)
* 11dd222 Test: refactor more tests (JPeer264)
* 92f4806 Style: if to better looking ternary operator (JPeer264)

3.0.0-alpha.10 - February, 21 2020

* 082fbcc Chore: update github repo (JPeer264)
* 4d387ab Docs: update badges in readme (JPeer264)
* 1a4ee5b Chore: update jest and outsource config (JPeer264)
* 4ec0ecf Chore: update husky and lint-staged (JPeer264)
* 9e13188 Chore: different method of publishing (JPeer264)

3.0.0-alpha.9 - February, 21 2020

* 6f025df Fix: add multiple classes divided by a space (closes #3) (JPeer264)

3.0.0-alpha.8 - February, 24 2019

* c7ce7c2 Fix: wrong progressbar behavior (closes #1) (JPeer264)
* c4883d9 Test: add more unit tests (JPeer264)

3.0.0-alpha.7 - February, 01 2019

* 5ca4b4d Chore: remove rollup docstrings, add dynamic style build (JPeer264)
* d9e426a Docs: update docs (JPeer264)
* eee1ba0 Feat: make options public so they can be changed (JPeer264)
* 295d4fa Style: remove one eslint error (JPeer264)
* 051f26f Chore: move tests, update rollup (JPeer264)
* a1f991f Test: add more tests (JPeer264)
* c8d33f5 Test: finish refactoring clear tests (JPeer264)

3.0.0-alpha.6 - February, 01 2019

* 6ad4726 Fix: optional types (JPeer264)

3.0.0-alpha.5 - February, 01 2019

* 7658574 Chore: add root dir to jest (JPeer264)
* 728b6af Chore: jest update testregex (JPeer264)
* d180e21 Test: fix tests (JPeer264)
* 432ebdf Chore: update rollup (JPeer264)
* 4472a88 Chore: update travis (JPeer264)
* c7c135b Test: outsource tests to have smaller test files (JPeer264)
* e40b76c Test: add first two tests (JPeer264)
* a482700 Chore: update test script (JPeer264)
* a8a9b00 Test: add basic jest tests (JPeer264)
* ce6e4a7 Chore: remove karma, prepare for jest (JPeer264)
* 5f95f0d Chore: remove and clearup dependencies (JPeer264)
* 2335d62 Chore: update little demo (JPeer264)
* fbae88d Chore: remove jshint (JPeer264)
* 7d51529 Refactor: remove nuget (JPeer264)

3.0.0-alpha.4 - January, 31 2019

* 44ad8b2 Chore: remove next tag (JPeer264)
* e364183 Docs: update readme (JPeer264)

3.0.0-alpha.3 - January, 31 2019

* d95a7c1 ocs: update docs (JPeer264)
* c0da425 Chore: update eslint dev dependencies (JPeer264)
* 923fb90 Refactor: uppercase Toastr (JPeer264)
* d11d136 Feat: export types (JPeer264)
* 67158b7 Feat: Toastr. change to class (JPeer264)
* f55780b Chore: add more types (JPeer264)

3.0.0-alpha.2 - January, 28 2019

* e202935 Chore: change typings location (JPeer264)

3.0.0-alpha.1 - January, 28 2019

* 6e5c230 Feat: add some types (JPeer264)

3.0.0-alpha.0 - January, 28 2019

* 7e3fc00 Chore: update tets (JPeer264)
* 0759723 Chore: change build to prepublish, works better with sgr (JPeer264)
* db865b7 Chore: update build (JPeer264)
* 8918efe Chore: update maintainer and package info (JPeer264)
* f497b96 Feat: add eslint airbnb styleguide (JPeer264)
* 21644bd Chore: change to typescript with no strict types (JPeer264)
* 744b65c Refactor: remove jquery (JPeer264)
* 50092cc Add yarn package manager to the list (#557) (iamfaizahmed123)
* 5084952 add badges and add jsdelivr (#554) (Stephen)
* aa1db85 added section for auto hiding (#576) (Marty Powell)
* 2fac397 Create LICENSE (Tim Ferrell)
* fc4ae18 Patching travis to use Node 6. (Timothy Ferrell)
* 4bddf04 Manually merging in @JLHwung 's fork to resolve CI issues. (Timothy Ferrell)
* 1ef00d7 Rev to 2.1.4. No changes except tag version. (Timothy Ferrell)
* 4ba6bb3 Add link to Rails gem (Alex Kalinin)
* 70dc5cf Fix headers in readme (Fernando van Loenhout)
* 3efb7d9 Update README.md (Rainer Eli)
* cdf2a6c CDN location changed (Alwin Kesler Pacheco)
* 28fb8e4 Fix missing import. Remove unused dev dependencies. (ricardo.brandao)
* 2601964 Adding tests for new aria functions. (Timothy Ferrell)
* a13f418 Updating bug fix notes for 2.1.3 (Timothy Ferrell)
* f6fb3e3 Updating version numbers. (Timothy Ferrell)
* fd9d772 Updating release notes for the 2.1.3 release. (Timothy Ferrell)
* 910b086 Removing unused functions and commented out code from the gulpfile. (Timothy Ferrell)
* ea6a5ed Add tests to RTL support (Ricardo Brandão)
* c43377c Add RTL support (closes #257 and closes #347) (Ricardo Brandão)
* 4d5bab6 close button now has line-height 1 to stay consistent over bootstrap versions (pickledyamsman)

demo/index.ts
import Toastr from '../src/Toastr';

const toastr = new Toastr();

let i = -1;
let toastIndex = 0;
let lastToast;

const getMessage = (): string => {
  const msgs = [
    'My name is Inigo Montoya. You killed my father. Prepare to die!',
    '<div><input class="input-small" value="textbox"/>&nbsp;<a href="http://johnpapa.net" target="_blank">This is a hyperlink</a></div><div><button type="button" id="okBtn" class="btn btn-primary">Close me</button><button type="button" id="surpriseBtn" class="btn" style="margin: 0 8px 0 8px">Surprise me</button></div>',
    'Are you the six fingered man?',
    'Inconceivable!',
    'I do not think that means what you think it means.',
    'Have fun storming the castle!',
  ];

  i += 1;

  if (i === msgs.length) {
    i = 0;
  }

  return msgs[i];
};

const getMessageWithClearButton = (msg = 'Clear itself?'): string => (
  `${msg}<br /><br /><button type="button" class="btn clear">Yes</button>`
);

const form = (): void => {
  const showToastForm = document.getElementById('toastForm');

  if (!showToastForm) {
    return;
  }

  showToastForm.addEventListener('submit', (e) => {
    e.preventDefault();

    const formData = new FormData(showToastForm as HTMLFormElement);

    let message = formData.get('message') as string;

    const shortCutFunction = formData.get('toastTypeGroup') as string;
    const title = formData.get('title');
    const showDuration = formData.get('showDuration');
    const hideDuration = formData.get('hideDuration');
    const timeOut = formData.get('timeOut');
    const extendedTimeOut = formData.get('extendedTimeOut');
    const showEasing = formData.get('showEasing');
    const hideEasing = formData.get('hideEasing');
    const showMethod = formData.get('showMethod');
    const hideMethod = formData.get('hideMethod');
    const addClear = formData.get('addClear');

    toastIndex += 1;

    toastr.options.closeButton = !!formData.get('closeButton');
    toastr.options.debug = !!formData.get('debugInfo');
    toastr.options.newestOnTop = !!formData.get('newestOnTop');
    toastr.options.progressBar = !!formData.get('progressBar');
    toastr.options.rtl = !!formData.get('rtl');
    toastr.options.positionClass = formData.get('positionGroup') as string || 'toast-top-right';
    toastr.options.preventDuplicates = !!formData.get('preventDuplicates');
    toastr.options.onclick = null;

    if (formData.get('addBehaviorOnToastClick')) {
      toastr.options.onclick = () => (
        // eslint-disable-next-line no-alert
        alert('You can perform some custom action after a toast goes away')
      );
    }

    if (formData.get('addBehaviorOnToastCloseClick')) {
      toastr.options.onCloseClick = () => (
        // eslint-disable-next-line no-alert
        alert('You can perform some custom action when the close button is clicked')
      );
    }

    if (showDuration) {
      toastr.options.showDuration = Number(showDuration);
    }

    if (hideDuration) {
      toastr.options.hideDuration = Number(hideDuration);
    }

    if (timeOut) {
      toastr.options.timeOut = addClear ? 0 : Number(timeOut);
    }

    if (extendedTimeOut) {
      toastr.options.extendedTimeOut = addClear ? 0 : Number(extendedTimeOut);
    }

    if (showEasing) {
      toastr.options.showEasing = showEasing as any;
    }

    if (hideEasing) {
      toastr.options.hideEasing = hideEasing as any;
    }

    if (showMethod) {
      toastr.options.showMethod = showMethod as any;
    }

    if (hideMethod) {
      toastr.options.hideMethod = hideMethod as any;
    }

    if (addClear) {
      message = getMessageWithClearButton(message);
      toastr.options.tapToDismiss = false;
    }

    if (!message) {
      message = getMessage();
    }

    const toastrOptionsEl = document.getElementById('toastrOptions');

    if (toastrOptionsEl) {
      toastrOptionsEl.innerHTML = `Command: toastr.${shortCutFunction}("${message}${title ? `, ${title}` : ''}")

toastr.options = ${JSON.stringify(toastr.options, null, 2)}
        `;
    }


    // Wire up an event handler to a button in the toast, if it exists
    lastToast = toastr[shortCutFunction](message, title) as HTMLElement;

    if (!lastToast) {
      return;
    }

    const lastOkBtn = lastToast.querySelector('#okBtn');
    const lastSurpriseBtn = lastToast.querySelector('#surpriseBtn');
    const lastClear = lastToast.querySelector('.clear');

    if (lastOkBtn) {
      lastOkBtn.addEventListener('click', () => {
        // eslint-disable-next-line no-alert
        alert(`you clicked me. i was toast #${toastIndex}. goodbye!`);

        lastToast.remove();
      });
    }

    if (lastSurpriseBtn) {
      lastSurpriseBtn.addEventListener('click', () => (
        // eslint-disable-next-line no-alert
        alert(`Surprise! you clicked me. i was toast #${toastIndex}. You could perform an action here.`)
      ));
    }

    if (lastClear) {
      lastClear.addEventListener('click', () => {
        toastr.clear(lastToast, { force: true });
      });
    }
  });
};

form();

const closeButtonEl = document.getElementById('closeButton') as HTMLInputElement;
const addBehaviorOnToastCloseClickEl = document.getElementById('addBehaviorOnToastCloseClick') as HTMLInputElement;
const clearLastToastEl = document.getElementById('clearlasttoast');
const clearAllToastsEl = document.getElementById('cleartoasts');

if (closeButtonEl && addBehaviorOnToastCloseClickEl) {
  closeButtonEl.addEventListener('click', () => {
    if (closeButtonEl.checked) {
      addBehaviorOnToastCloseClickEl.removeAttribute('disabled');
    } else {
      addBehaviorOnToastCloseClickEl.setAttribute('disabled', '');
      addBehaviorOnToastCloseClickEl.checked = false;
    }
  });
}

if (clearLastToastEl) {
  clearLastToastEl.addEventListener('click', () => {
    toastr.clear(lastToast);
  });
}

if (clearAllToastsEl) {
  clearAllToastsEl.addEventListener('click', () => {
    toastr.clear();
  });
}


src/Toastr.ts
import merge from 'lodash/merge';

import ProgressBar from './additions/ProgressBar';
import './toastr.scss';
import { version } from '../package.json';
import addClasses from './helpers/addClasses';

type Required<T> = {
  [P in keyof T]-?: T[P];
}

export type ToastType = {
  info?: string;
  error?: string;
  warning?: string;
  success?: string;
};

export type RequiredToastType = Required<ToastType>;

export type ToastrOptions<T = ToastType> = {
  tapToDismiss?: boolean;
  toastClass?: string | string[];
  containerId?: string;
  debug?: boolean;

  showMethod?: 'fadeIn' | 'slideDown' | 'show';
  showDuration?: number;
  showEasing?: 'swing' | 'linear';
  onShown?: () => void;
  hideMethod?: 'fadeOut';
  hideDuration?: number;
  hideEasing?: 'swing';
  onHidden?: () => void;
  closeMethod?: boolean;
  closeDuration?: number | false;
  closeEasing?: boolean;
  closeOnHover?: boolean;

  extendedTimeOut?: number;
  iconClasses?: T;
  iconClass?: string | string[];
  positionClass?: string | string[];
  timeOut?: number; // Set timeOut and extendedTimeOut to 0 to make it sticky
  titleClass?: string | string[];
  messageClass?: string | string[];
  escapeHtml?: boolean;
  target?: string;
  closeHtml?: string;
  closeClass?: string | string[];
  newestOnTop?: boolean;
  preventDuplicates?: boolean;
  progressBar?: boolean;
  progressClass?: string | string[];
  onclick?: (event: MouseEvent) => void;

  onCloseClick?: (event: Event) => void;
  closeButton?: boolean;
  rtl?: boolean;
}

export type NotifyMap = {
  type: string;
  optionsOverride?: ToastrOptions;
  iconClass: string;
  title?: string;
  message?: string;
}

class Toastr {
  private listener: any;

  private toastId = 0;

  private previousToast: string | null = null;

  private toastType: RequiredToastType = {
    info: 'info',
    error: 'error',
    warning: 'warning',
    success: 'success',
  };

  private version = version;

  public options: Required<ToastrOptions<RequiredToastType>> = {
    tapToDismiss: true,
    toastClass: 'toast',
    containerId: 'toast-container',
    debug: false,

    showMethod: 'fadeIn', // fadeIn, slideDown, and show are built into jQuery
    showDuration: 300,
    showEasing: 'swing', // swing and linear are built into jQuery
    onShown: () => { },
    hideMethod: 'fadeOut',
    hideDuration: 1000,
    hideEasing: 'swing',
    onHidden: () => { },
    closeMethod: false,
    closeDuration: false,
    closeEasing: false,
    closeOnHover: true,

    extendedTimeOut: 1000,
    iconClasses: {
      error: 'toast-error',
      info: 'toast-info',
      success: 'toast-success',
      warning: 'toast-warning',
    },
    iconClass: 'toast-info',
    positionClass: 'toast-top-right',
    timeOut: 5000, // Set timeOut and extendedTimeOut to 0 to make it sticky
    titleClass: 'toast-title',
    messageClass: 'toast-message',
    escapeHtml: false,
    target: 'body',
    closeHtml: '<button type="button">&times;</button>',
    closeClass: 'toast-close-button',
    newestOnTop: true,
    preventDuplicates: false,
    progressBar: false,
    progressClass: 'toast-progress',
    rtl: false,

    onCloseClick: () => { },
    closeButton: false,

    onclick: () => { },
  };

  public $container: HTMLElement = document.createElement('div');

  public constructor(options?: ToastrOptions) {
    this.options = merge({}, this.options, options);

    this.createContainer();
  }

  public createContainer(): HTMLElement {
    this.$container = document.createElement('div');

    this.$container.setAttribute('id', this.options.containerId);
    addClasses(this.$container, this.options.positionClass);

    const target = document.getElementsByTagName(this.options.target);

    if (target && target[0]) {
      target[0].appendChild(this.$container);
    }

    return this.$container;
  }

  public getContainer(options: Partial<ToastrOptions> = this.options, create = false): HTMLElement {
    const $container = document.getElementById(options.containerId || '');

    if ($container) {
      this.$container = $container;

      return this.$container;
    }

    if (create) {
      this.$container = this.createContainer();
    }

    return this.$container;
  }

  public error(
    message?: string,
    title?: string,
    optionsOverride?: ToastrOptions,
  ): HTMLElement | null {
    return this.notify({
      type: this.toastType.error,
      iconClass: this.options.iconClasses.error,
      message,
      optionsOverride,
      title,
    });
  }

  public warning(
    message?: string,
    title?: string,
    optionsOverride?: ToastrOptions,
  ): HTMLElement | null {
    return this.notify({
      type: this.toastType.warning,
      iconClass: this.options.iconClasses.warning,
      message,
      optionsOverride,
      title,
    });
  }

  public success(
    message?: string,
    title?: string,
    optionsOverride?: ToastrOptions,
  ): HTMLElement | null {
    return this.notify({
      type: this.toastType.success,
      iconClass: this.options.iconClasses.success,
      message,
      optionsOverride,
      title,
    });
  }

  public info(
    message?: string,
    title?: string,
    optionsOverride?: ToastrOptions,
  ): HTMLElement | null {
    return this.notify({
      type: this.toastType.info,
      iconClass: this.options.iconClasses.info,
      message,
      optionsOverride,
      title,
    });
  }

  public subscribe(callback: (response: Toastr) => void): void {
    this.listener = callback;
  }

  public publish(args: Toastr): void {
    if (!this.listener) {
      return;
    }

    this.listener(args);
  }

  public clear(toastElement?: HTMLElement | null, clearOptions: { force?: boolean } = {}) {
    if (!this.$container) {
      this.getContainer(this.options);
    }

    if (!this.clearToast(toastElement, this.options, clearOptions)) {
      this.clearContainer(this.options);
    }
  }

  public remove(toastElement?: HTMLElement | null) {
    if (!this.$container) {
      this.getContainer(this.options);
    }

    if (!this.$container) {
      return;
    }

    if (toastElement && toastElement !== document.activeElement) {
      this.removeToast(toastElement);

      return;
    }

    if (!this.$container.hasChildNodes()) {
      const parentNode = this.$container.parentElement;

      if (parentNode) {
        parentNode.removeChild(this.$container);
      }
    }
  }

  public removeToast(toastElement: HTMLElement) {
    if (!this.$container) {
      this.getContainer();
    }

    if (!this.$container || !toastElement.parentNode) {
      return;
    }

    // todo set after visible state
    // as this will be a transition of css
    toastElement.parentNode.removeChild(toastElement);
    // check if visible
    if (toastElement.offsetWidth > 0 && toastElement.offsetHeight > 0) {
      return;
    }

    // todo check if null makes sense
    // toastElement = null;

    if (!this.$container.hasChildNodes()) {
      if (this.$container.parentNode) {
        this.$container.parentNode.removeChild(this.$container);
      }

      this.previousToast = null;
    }
  }

  private clearContainer(options: Partial<ToastrOptions> = this.options) {
    if (!this.$container) {
      return;
    }

    const toastsToClear = Array.from(this.$container.childNodes) as HTMLElement[];

    for (let i = toastsToClear.length - 1; i >= 0; i -= 1) {
      this.clearToast(toastsToClear[i], options);
    }
  }

  private clearToast(
    toastElement?: HTMLElement | null,
    // eslint-disable-next-line no-unused-vars
    options: Partial<ToastrOptions> = this.options,
    clearOptions: { force?: boolean } = {},
  ): boolean {
    if (!toastElement) {
      return false;
    }

    const force = clearOptions.force || false;

    if (toastElement && (force || toastElement !== document.activeElement)) {
      // todo hide effect
      this.removeToast(toastElement);
      // toastElement[options.hideMethod]({
      //     duration: options.hideDuration,
      //     easing: options.hideEasing,
      //     complete: function () { removeToast(toastElement); }
      // });
      return true;
    }

    return false;
  }

  private notify(map: NotifyMap): HTMLElement | null {
    let { options } = this;
    let iconClass = map.iconClass || this.options.iconClass;

    const shouldExit = (opts: ToastrOptions, exitMap: NotifyMap): boolean => {
      if (opts.preventDuplicates) {
        if (exitMap.message === this.previousToast) {
          return true;
        }

        this.previousToast = exitMap.message || '';
      }
      return false;
    };

    if (typeof map.optionsOverride !== 'undefined') {
      options = merge({}, options, map.optionsOverride);
      iconClass = map.optionsOverride.iconClass || iconClass;
    }

    if (shouldExit(options, map)) {
      return null;
    }

    this.toastId += 1;

    this.$container = this.getContainer(options, true);

    let intervalId: NodeJS.Timeout | null = null;
    let progressBar: null | ProgressBar = null;
    const toastElement = document.createElement('div');
    const $titleElement = document.createElement('div');
    const $messageElement = document.createElement('div');
    const createdElement = document.createElement('div');
    createdElement.innerHTML = options.closeHtml.trim();
    const closeElement = createdElement.firstChild as HTMLElement | null;

    const response: any = {
      toastId: this.toastId,
      state: 'visible',
      startTime: new Date(),
      endTime: undefined,
      options,
      map,
    };

    const hideToast = (override: any = null): void => {
      // const method = override && this.options.closeMethod !== false
      //   ? this.options.closeMethod
      //   : this.options.hideMethod;

      // const duration = override && this.options.closeDuration !== false
      //   ? this.options.closeDuration
      //   : this.options.hideDuration;

      // const easing = override && this.options.closeEasing !== false
      //   ? this.options.closeEasing
      //   : this.options.hideEasing;

      if (toastElement === document.activeElement && !override) {
        return;
      }

      if (progressBar) {
        progressBar.stop();
      }

      // todo fade out toast
      this.removeToast(toastElement);

      if (intervalId) {
        clearTimeout(intervalId);
      }

      if (options.onHidden && response.state !== 'hidden') {
        options.onHidden();
      }

      response.state = 'hidden';
      response.endTime = new Date();
      this.publish(response);

      // return toastElement[method]({
      //     duration: duration,
      //     easing: easing,
      // });
    };

    const escapeHtml = (source: string | null): string => {
      const newSource = source !== null ? source : '';

      return newSource
        .replace(/&/g, '&amp;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#39;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;');
    };

    const setAria = (): void => {
      let ariaValue = '';

      switch (iconClass) {
        case 'toast-success':
        case 'toast-info':
          ariaValue = 'polite';

          break;

        default:
          ariaValue = 'assertive';
      }

      toastElement.setAttribute('aria-live', ariaValue);
    };

    const delayedHideToast = (): void => {
      if (options.timeOut > 0 || options.extendedTimeOut > 0) {
        intervalId = setTimeout(hideToast, options.extendedTimeOut);

        if (progressBar) {
          progressBar.reset(options.extendedTimeOut);
          progressBar.start();
        }
      }
    };

    const stickAround = (): void => {
      if (intervalId) {
        clearTimeout(intervalId);
      }

      if (progressBar) {
        progressBar.stop();
      }
      // todo
      // toastElement.stop(true, true)[options.showMethod](
      //     {duration: options.showDuration, easing: options.showEasing}
      // );
    };

    const handleEvents = (): void => {
      if (options.closeOnHover) {
        toastElement.addEventListener('mouseover', () => stickAround());
        toastElement.addEventListener('mouseout', () => delayedHideToast());
      }

      if (!options.onclick && options.tapToDismiss) {
        toastElement.addEventListener('click', hideToast);
      }

      if (options.closeButton && closeElement) {
        closeElement.addEventListener('click', (event) => {
          if (event.stopPropagation) {
            event.stopPropagation();
          } else if (event.cancelBubble !== undefined && event.cancelBubble !== true) {
            // eslint-disable-next-line no-param-reassign
            event.cancelBubble = true;
          }

          if (options.onCloseClick) {
            options.onCloseClick(event);
          }

          hideToast(true);
        });
      }

      if (options.onclick) {
        toastElement.addEventListener('click', (event) => {
          // ts needs another check here
          if (options.onclick) {
            options.onclick(event);
          }

          hideToast();
        });
      }
    };

    const setTitle = (): void => {
      if (map.title) {
        let suffix = map.title;
        if (options.escapeHtml) {
          suffix = escapeHtml(map.title);
        }
        $titleElement.innerHTML = suffix;
        addClasses($titleElement, options.titleClass);
        toastElement.appendChild($titleElement);
      }
    };

    const setMessage = (): void => {
      if (map.message) {
        let suffix = map.message;

        if (options.escapeHtml) {
          suffix = escapeHtml(map.message);
        }

        $messageElement.innerHTML = suffix;
        addClasses($messageElement, options.messageClass);
        toastElement.appendChild($messageElement);
      }
    };

    const setCloseButton = (): void => {
      if (options.closeButton && closeElement) {
        addClasses(closeElement, options.closeClass);
        closeElement.setAttribute('role', 'button');
        toastElement.insertBefore(closeElement, toastElement.firstChild);
      }
    };

    const setProgressBar = (): void => {
      if (options.progressBar) {
        progressBar = new ProgressBar(toastElement, options.progressClass);
      }
    };

    const setRTL = (): void => {
      if (options.rtl) {
        addClasses(toastElement, 'rtl');
      }
    };

    const setIcon = (): void => {
      if (iconClass) {
        addClasses(toastElement, options.toastClass, iconClass);
      }
    };

    const setSequence = (): void => {
      if (options.newestOnTop) {
        this.$container.insertBefore(toastElement, this.$container.firstChild);
      } else {
        this.$container.appendChild(toastElement);
      }
    };

    const displayToast = (): void => {
      // todo hide toast
      // toastElement.hide();

      // todo fade out toast
      if (options.onShown) {
        options.onShown();
      }
      // toastElement[options.showMethod](
      // eslint-disable-next-line
      //     {duration: options.showDuration, easing: options.showEasing, complete: options.onShown}
      // );

      if (options.timeOut > 0) {
        intervalId = setTimeout(hideToast, options.timeOut);

        if (progressBar) {
          progressBar.reset(options.timeOut);
          progressBar.start();
        }
      }
    };

    const personalizeToast = (): void => {
      setIcon();
      setTitle();
      setMessage();
      setCloseButton();
      setProgressBar();
      setRTL();
      setSequence();
      setAria();
    };

    personalizeToast();

    displayToast();

    handleEvents();

    this.publish(response);

    if (options.debug && console) {
      // eslint-disable-next-line no-console
      console.log(response);
    }

    return toastElement;
  }
}

export default Toastr;
